File: model-client/src/main/java/org/modelix/model/lazy/BulkQuery.java
package org.modelix.model.lazy;

import io.vavr.Tuple;
import io.vavr.Tuple3;
import org.apache.commons.lang3.mutable.MutableInt;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

/**
 * Not thread safe
 */
public class BulkQuery implements IBulkQuery {

  private IDeserializingKeyValueStore store;
  private List<Tuple3<String, Function<String, ?>, Consumer<Object>>> queue = new ArrayList<>();
  private boolean processing = false;

  public BulkQuery(IDeserializingKeyValueStore store) {
    this.store = store;
  }

  protected Map<String, Object> executeBulkQuery(Iterable<String> keys, final Map<String, Function<String, ?>> deserializers) {
    Iterable<Object> values = store.getAll(keys, (key, serialized) -> deserializers.get(key).apply(serialized));
    Map<String, Object> result = new HashMap<>();
    {
      Iterator<String> key_it = keys.iterator();
      Iterator<Object> value_it = values.iterator();
      String key_var;
      Object value_var;
      while (key_it.hasNext() && value_it.hasNext()) {
        key_var = key_it.next();
        value_var = value_it.next();
        result.put(key_var, value_var);
      }
    }
    return result;
  }

  public void query(String key, Function<String, ?> deserializer, Consumer<Object> callback) {
    queue.add(Tuple.of(key, deserializer, callback));
  }

  public <T> IBulkQuery.Value<T> get(String hash, Function<String, T> deserializer) {
    final Value<T> result = new Value<>();
    query(hash, deserializer, value -> result.success((T) value));
    return result;
  }

  public <T> IBulkQuery.Value<T> constant(T value) {
    return new Value<T>(value);
  }

  public void process() {
    if (processing) {
      throw new RuntimeException("Already processing");
    }
    processing = true;
    try {
      while (!queue.isEmpty()) {
        final List<Tuple3<String, Function<String, ?>, Consumer<Object>>> currentRequests = queue;
        queue = new ArrayList<>();

        Map<String, Function<String, ?>> deserializers = new HashMap<>();
        for (Tuple3<String, Function<String, ?>, Consumer<Object>> request : currentRequests) {
          deserializers.put(request._1(), request._2());
        }

        final Map<String, Object> entries = executeBulkQuery(
                currentRequests.stream()
                        .map(Tuple3::_1)
                        .distinct()
                        .collect(Collectors.toList()),
                deserializers);
        for (Tuple3<String, Function<String, ?>, Consumer<Object>> request : currentRequests) {
          request._3().accept(entries.get(request._1()));
        }
      }
    } finally {
      processing = false;
    }
  }

  public <I, O> IBulkQuery.Value<List<O>> map(Iterable<I> input_, Function<I, IBulkQuery.Value<O>> f) {
    List<I> input = StreamSupport.stream(input_.spliterator(), false).collect(Collectors.toList());
    if (input.isEmpty()) {
      return constant(Collections.<O>emptyList());
    }
    final Object[] output = new Object[input.size()];
    final boolean[] done = new boolean[input.size()];
    final MutableInt remaining = new MutableInt(input.size());
    final Value<List<O>> result = new Value<>();
    for (int i_ = 0; i_ < input.size(); i_++) {
      final int i = i_;
      f.apply(input.get(i)).onSuccess(new Consumer<O>() {
        public void accept(O value) {
          if (done[i]) {
            return;
          }
          output[i] = value;
          done[i] = true;
          remaining.decrement();
          if (remaining.intValue() == 0) {
            result.success(Arrays.stream(output).map(e -> (O) e).collect(Collectors.toList()));
          }
        }
      });
    }
    return result;
  }

  public class Value<T> implements IBulkQuery.Value<T> {
    private List<Consumer<T>> handlers = new ArrayList<>();
    private T value;
    private boolean done = false;

    public Value() {
    }

    public Value(T value) {
      this.value = value;
      done = true;
    }

    public synchronized void success(T value) {
      if (done) {
        throw new IllegalStateException("Value is already set");
      } else {
        this.value = value;
        this.done = true;
        for (Consumer<T> handler : handlers) {
          handler.accept(value);
        }
        handlers = null;
      }
    }

    public synchronized void onSuccess(Consumer<T> handler) {
      if (done) {
        handler.accept(value);
      } else {
        handlers.add(handler);
      }
    }

    public T execute() {
      process();
      if (!(done)) {
        throw new RuntimeException("No value received");
      }
      return value;
    }

    public <R> IBulkQuery.Value<R> map(final Function<T, R> handler) {
      final Value<R> result = new Value<>();
      onSuccess(v -> result.success(handler.apply(v)));
      return result;
    }

    public <R> IBulkQuery.Value<R> mapBulk(final Function<T, IBulkQuery.Value<R>> handler) {
      final Value<R> result = new Value<>();
      onSuccess(v -> handler.apply(v).onSuccess(result::success));
      return result;
    }
  }

}

File: model-client/src/main/java/org/modelix/model/lazy/BulkQuery.kt
package org.modelix.model.lazy

import io.vavr.Tuple
import io.vavr.Tuple3
import org.apache.commons.lang3.mutable.MutableInt
import java.util.*
import java.util.function.Consumer
import java.util.function.Function
import java.util.stream.Collectors
import java.util.stream.StreamSupport

/**
 * Not thread safe
 */
class BulkQuery(private val store: IDeserializingKeyValueStore) : IBulkQuery {
    private var queue: MutableList<Tuple3<String, Function<String, *>, Consumer<Any?>>> = ArrayList()
    private var processing = false
    protected fun executeBulkQuery(keys: Iterable<String>, deserializers: Map<String, Function<String, *>>): Map<String, Any> {
        val values = store.getAll(keys) { key: String?, serialized: String? -> deserializers[key]!!.apply(serialized!!) }
        val result: MutableMap<String, Any> = HashMap()
        run {
            val key_it = keys.iterator()
            val value_it: Iterator<Any> = values.iterator()
            var key_var: String
            var value_var: Any
            while (key_it.hasNext() && value_it.hasNext()) {
                key_var = key_it.next()
                value_var = value_it.next()
                result[key_var] = value_var
            }
        }
        return result
    }

    fun query(key: String, deserializer: Function<String, *>, callback: Consumer<Any?>) {
        queue.add(Tuple.of<String, Function<String, *>, Consumer<Any?>>(key, deserializer, callback))
    }

    override fun <T> get(hash: String, deserializer: Function<String, T>): IBulkQuery.Value<T> {
        val result = Value<T>()
        query(hash, deserializer, Consumer { value: Any? -> result.success(value as T) })
        return result
    }

    override fun <T> constant(value: T): IBulkQuery.Value<T> {
        return Value(value)
    }

    fun process() {
        if (processing) {
            throw RuntimeException("Already processing")
        }
        processing = true
        try {
            while (!queue.isEmpty()) {
                val currentRequests: List<Tuple3<String, Function<String, *>, Consumer<Any?>>> = queue
                queue = ArrayList()
                val deserializers: MutableMap<String, Function<String, *>> = HashMap()
                for (request in currentRequests) {
                    deserializers[request._1()] = request._2()
                }
                val entries = executeBulkQuery(
                        currentRequests.stream()
                                .map { obj: Tuple3<String, Function<String, *>, Consumer<Any?>> -> obj._1() }
                                .distinct()
                                .collect(Collectors.toList()),
                        deserializers)
                for (request in currentRequests) {
                    request._3().accept(entries[request._1()])
                }
            }
        } finally {
            processing = false
        }
    }

    override fun <I, O> map(input_: Iterable<I>, f: Function<I, IBulkQuery.Value<O>>): IBulkQuery.Value<List<O>> {
        val input = StreamSupport.stream(input_.spliterator(), false).collect(Collectors.toList())
        if (input.isEmpty()) {
            return constant(emptyList())
        }
        val output = arrayOfNulls<Any>(input.size)
        val done = BooleanArray(input.size)
        val remaining = MutableInt(input.size)
        val result = Value<List<O>>()
        for (i_ in input.indices) {
            f.apply(input[i_]).onSuccess(Consumer { value ->
                if (done[i_]) {
                    return@Consumer
                }
                output[i_] = value
                done[i_] = true
                remaining.decrement()
                if (remaining.toInt() == 0) {
                    result.success(Arrays.stream(output).map { e: Any? -> e as O }.collect(Collectors.toList()))
                }
            })
        }
        return result
    }

    inner class Value<T> : IBulkQuery.Value<T> {
        private var handlers: MutableList<Consumer<T?>>? = ArrayList()
        private var value: T? = null
        private var done = false

        constructor() {}
        constructor(value: T) {
            this.value = value
            done = true
        }

        @Synchronized
        fun success(value: T) {
            check(!done) { "Value is already set" }
            this.value = value
            done = true
            for (handler in handlers!!) {
                handler.accept(value)
            }
            handlers = null
        }

        @Synchronized
        override fun onSuccess(handler: Consumer<T?>) {
            if (done) {
                handler.accept(value)
            } else {
                handlers!!.add(handler)
            }
        }

        override fun execute(): T {
            process()
            if (!done) {
                throw RuntimeException("No value received")
            }
            return value!!
        }

        override fun <R> map(handler: Function<T, R>): IBulkQuery.Value<R> {
            val result = Value<R>()
            onSuccess { v: T? -> result.success(handler.apply(v!!)) }
            return result
        }

        override fun <R> mapBulk(handler: Function<T, IBulkQuery.Value<R>>): IBulkQuery.Value<R> {
            val result = Value<R>()
            onSuccess { v: T? -> handler.apply(v!!).onSuccess { value: R -> result.success(value) } }
            return result
        }
    }

}